### PR TITLE
Make pushing idempotent, and after-write job enqueuing atomic with making the version indexed

### PR DESCRIPTION
**Pull request by segiddins**
Goal here is to avoid issues where the connection drops _after_ the gem has been written to storage, but _before_ all the after-write jobs get enqueued.

Fix is to delay making the gem visible (indexed=true) until the transaction where those jobs get enqueued (which is then atomic, since they get enqueued to the same postgres database)

### CODE REVIEW

File: app/jobs/after_version_write_job.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+class AfterVersionWriteJob < ApplicationJob
+  queue_as :default
+
+  def perform(version:)
+    version.transaction do
+      rubygem = version.rubygem
+      version.rubygem.push_notifiable_owners.each do |notified_user|
+        Mailer.gem_pushed(owner, version.id, notified_user.id).deliver_later
```
Comment (simi) [inline app/jobs/after_version_write_job.rb:8]: There's chance `owner` is `nil` here. Is that ok? Mail would fail, but in the background.

Comment (segiddins) [reply app/jobs/after_version_write_job.rb:8]: fixed!

**Comment by segiddins** [comment]
@simi I think this is ready for you to take a look

File: app/jobs/after_version_write_job.rb (hunk 1 above)
Comment (simi) [reply app/jobs/after_version_write_job.rb:8]: Is it? :thinking: If I understand it well, owner still could be `nil`.

File: lib/certificate_chain_serializer.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@ class CertificateChainSerializer
 
   def self.load(chain)
     return [] unless chain
-    chain.scan(PATTERN).map do |cert|
+    chain.scan(PATTERN).map! do |cert|
```
Comment (simi) [inline lib/certificate_chain_serializer.rb:6]: Is this change related to the functionality?

File: test/unit/mailer_test.rb [hunk 3]
```diff
@@ -4,6 +4,10 @@ class MailerTest < ActionMailer::TestCase
   MIN_DOWNLOADS_FOR_MFA_RECOMMENDATION_POLICY = 165_000_000
   MIN_DOWNLOADS_FOR_MFA_REQUIRED_POLICY = 180_000_000
 
+  setup do
+    TOPLEVEL_BINDING.receiver.stubs(:mx_exists?).returns(true)
```
Comment (simi) [inline test/unit/mailer_test.rb:8]: Any details on this change?

Comment (segiddins) [reply test/unit/mailer_test.rb:8]: Needed to test the PR without internet access

File: lib/certificate_chain_serializer.rb (hunk 2 above)
Comment (segiddins) [reply lib/certificate_chain_serializer.rb:6]: No functional change

File: app/jobs/after_version_write_job.rb (hunk 1 above)
Comment (segiddins) [reply app/jobs/after_version_write_job.rb:8]: Mailer now handles nil

**Review by simi — APPROVED** [review]

**Review by simi — APPROVED** [review]
